### ISSUE TITLE
qBittorrent support

### ISSUE THREAD

**Issue by mmgoodnow** [mmgoodnow/cross-seed#49]
- direct client injection

#48

**Comment by Drazzilb08** [comment]
Let me know if there is anything I can do to help with testing.

**Comment by mmgoodnow** [comment]
@Drazzilb08 @Visorask `v2.3.1` has experimental support for direct torrent injection into qBittorrent. 

I would greatly appreciate some testing!

Here's how to set it up with a config file. 
http://github.com/mmgoodnow/cross-seed/blob/c426af9b3f38e36eb8ea9c9d08a54834eb30ae55/src/config.template.js#L44-L48

You can also use the `--qbittorrent-url` command-line flag.

**Comment by Drazzilb08** [comment]
So I got this to run this morning. I had some trouble with the config file as it does not seem to be reading the URL for qbit from the config file as I kept getting the error stating that a url is needed if using the inject param. So I ended up having to use the --qbittorrent-url in the post arguments section. I've ran it through and got a few matches for new torrents. It injects them in an errored state. (mainly because it does not define the categories that the parent torrent is from). The cross-seed category is a nice touch though.

**Comment by mmgoodnow** [comment]
Maybe a bit of a dumb question, but I'm a little confused on what makes categories different from tags?

**Comment by mmgoodnow** [comment]
And re: the save location, it should be getting that correct regardless of tag. It determines the save location by querying qbit for the save location of the original torrent. So just to confirm, it's definitely getting that wrong?

**Comment by Drazzilb08** [comment]
Yes. They all were added in an error state without a category set.

**Comment by Drazzilb08** [comment]
> Maybe a bit of a dumb question, but I'm a little confused on what makes categories different from tags?

But a category is associated to a folder EG:
`Movies` will be associated with `/data/.torrents/movies` and `Completed-Movies` would be associated with `/data/.torrents/completed-movies` 
You can set the folder to whatever you want and the category to what ever as well.

**Comment by mmgoodnow** [comment]
Hmm, good to know. I will take another look at that. 

Regarding your question about multiple clients, probably not in the foreseeable future. Feel free to open a feature request issue for it but it will require a fair amount of work/rearchitecture with no direct payoff for me, so it'll have to be a pretty popular feature request for me to take it on. Of course, I always welcome PRs.

**Comment by mmgoodnow** [comment]
Daemon mode *might* work. The real reason it's "rtorrent only" is that we have to get a list of infohashes to exclude from matching (because they would be duplicates). We also have to do this fresh every time we receive a request to make sure the list is up to date. 

Reading every torrent file into memory just to get its infohash is pretty slow (you've probably noticed how slow it is during the startup of `cross-seed search`) so we rely on the naming scheme of the rtorrent session directory, in which all torrent files are named `<infohash>.torrent`. So all we have to do is read all the file names in the session directory, and we have the list of infohashes. 

TLDR if qBittorrent names the torrent files in its session directory by infohash, and you also use the session dir as your torrent dir for cross-seed, then it should work. 

I have some work planned in #53 and #54 to stop relying on that behavior of rtorrent, and at that point daemon mode should "Just Work" for all existing and future clients.

**Comment by Drazzilb08** [comment]
Sounds good. Sticking with search for the mean time.

**Comment by mmgoodnow** [comment]
You can find out why things aren't being selected for searching with verbose logging

**Comment by mmgoodnow** [comment]
It does not. It was added before exclude-recent-search so the naming is a little suboptimal. But searchAll only prevents things from being filtered out based on their contents.

**Comment by mmgoodnow** [comment]
I tested it out more thoroughly and it's still working fine for me. I think there might be something going on with automatic torrent management. Can you look at the save paths of the cross-seeds? How are they different from the save paths of the original torrents?

**Comment by Drazzilb08** [comment]
<img width="580" alt="Screen Shot 2021-04-02 at 7 47 55 AM" src="https://user-images.githubusercontent.com/65743746/113426167-eb26f900-9387-11eb-84a4-65cb7799ba6d.png">
Here you can see these are all the same releases but two are errored with the save paths wrong. Automatic torrent management is selected as well on the cross-seeded entries.

**Comment by mmgoodnow** [comment]
I see you've got a `completed-movies` folder. What is the trigger that moves things to that folder?

**Comment by Drazzilb08** [comment]
Radarr is what triggers as soon as the video file is imported to my plex dir (via hard link) radarr will move the category over to completed

**Comment by mmgoodnow** [comment]
I haven't forgotten about this! Hoping to dig back in later this week.

**Comment by mmgoodnow** [comment]
Just released [`v2.3.3`](https://github.com/mmgoodnow/cross-seed/releases/tag/v2.3.3). 
Changes: 
- allow file configuration
- set category on injection
- skip injection when the existing torrent isn't complete yet

Give it a whirl!

**Comment by Drazzilb08** [comment]
Will do. :)

**Comment by mmgoodnow** [comment]
Any updates here? No rush if not!

**Comment by Drazzilb08** [comment]
Yeah, been running it for well over a week even with my massive collection of torrents everything is running very well. Everything has been completely automated. 
For me at least I can consider this one done.

**Comment by mmgoodnow** [comment]
Nice! Very glad to hear that. Per chance do you have any torrents without a category that you know it worked on? 

Alternatively, @Visorask do you know how a category-less torrent would show up in the API and if I'd need to omit category from the list of properties when I add the new torrents?

**Comment by bobokun** [comment]
> Alternatively, @Visorask do you know how a category-less torrent would show up in the API and if I'd need to omit category from the list of properties when I add the new torrents?

Hey I'm also one of the developers working on [qbit_manage ](https://github.com/StuffAnThings/qbit_manage). If a torrent doesn't have a category it would show up as `''`
Ex. `torrent.category == ''`

**Comment by mmgoodnow** [comment]
And if I pass in `''` as the category to the `add torrent` endpoint, that will be treated as "no category"?

**Comment by bobokun** [comment]
Yes that's correct

**Comment by mmgoodnow** [comment]
Great! I'll close this one out then. Thanks for all your help folks!
